garden_ai/metadata.py

```diff
@@ -0,0 +1,224 @@
+    new_cell = nbformat.v4.new_code_cell(notebook_metadata)
+    new_cell["metadata"] = {
+        "tags": ["garden_metadata_cell"],
+        "garden_metadata": {"NOTEBOOK_BASE_IMAGE_URI": None},
+    }
+    del new_cell["id"]
+    ntbk.cells.insert(0, new_cell)
+
+    # Write updated notebook data to file
+    nbformat.write(ntbk, notebook_path, version=nbformat.NO_CONVERT)
+
+
+def set_notebook_metadata(
+    notebook_path: Path,
+    base_image_name: Optional[str],
+    base_image_uri: str,
+    requirements_data: Optional[dict],
+):
+    ntbk = _read_notebook(notebook_path)
+
+    # Find cell with 'garden_metadata_cell' tag
+    for cell in ntbk.cells:
+        cell_tags = cell.get("metadata", {}).get("tags", [])
+        if "garden_metadata_cell" in cell_tags:
+            # Replace old cell source with new metadata
+            metadata_string = json.dumps(
+                {
+                    "NOTEBOOK_BASE_IMAGE_NAME": base_image_name,
+                    "NOTEBOOK_REQUIREMENTS": requirements_data,
+                },
+                indent=2,
+            )
+
+            cell["source"] = NOTEBOOK_METADATA_CELL_TEMPLATE.substitute(
+                metadata=metadata_string
+            )
+
+            # Add hidden metadata
+            cell["metadata"]["garden_metadata"][
+                "NOTEBOOK_BASE_IMAGE_URI"
+            ] = base_image_uri
+
+    # Write updated notebook data to file
+    nbformat.write(ntbk, notebook_path, version=nbformat.NO_CONVERT)
+
+
+def get_notebook_metadata(notebook_path: Path) -> dict:
+    notebook_metadata: dict = {}
+    notebook_metadata["notebook_image_name"] = None
+    notebook_metadata["notebook_requirements"] = None
+    notebook_metadata_cell_source = None
+    notebook_metadata_string = None
+
+    ntbk = _read_notebook(notebook_path)
+
+    # Find cell with 'garden_metadata_cell' tag and get source
+    for cell in ntbk.cells:
+        cell_tags = cell.get("metadata", {}).get("tags", [])
+        if "garden_metadata_cell" in cell_tags:
+            # Grab cell source and hidden metadata
+            notebook_metadata_cell_source = cell.get("source", None)
+            notebook_hidden_metadata = cell.get("metadata", {}).get(
+                "garden_metadata", {}
+            )
+            break
+
+    # Return empty notebook metadata dict if was unable to find cell source
+    if not notebook_metadata_cell_source:
+        typer.echo("Unable to find garden metadata cell.")
+        return notebook_metadata
+
+    # Grab the part of cell source with the metadata dict in it
+    clean_source = notebook_metadata_cell_source.replace("\n", "")
+    regex_get_metadata = r"^(.*?)\"\"\"(.*?)NOTEBOOK_METADATA(.*?)=(.*?)\"\"\"(.*?)$"
+
+    metadata_match = re.match(regex_get_metadata, clean_source)
+    if metadata_match and len(metadata_match.groups()) == 5:
+        notebook_metadata_string = metadata_match[4]
+
+    if notebook_metadata_string:
```

Comment:
Feels a little weird that we set `notebook_metadata_string` only in the above `if` block but then make a separate `if` block right after.